File: forge/src/main/java/dev/kineticcat/hexportation/forge/transfer/FluidVariantImpl.java
package dev.kineticcat.hexportation.forge.transfer;

import dev.kineticcat.hexportation.api.transfer.FluidVariant;
import net.minecraft.nbt.CompoundTag;
import net.minecraft.world.level.material.Fluid;
import net.minecraftforge.fluids.FluidStack;

/**
 * Forge implementation using singleton injection pattern.
 */
public class FluidVariantImpl extends FluidVariant.Implementation {
    
    public static final FluidVariantImpl INSTANCE = new FluidVariantImpl();
    
    @Override
    public FluidVariant of(Fluid fluid) {
        return new FluidStackWrapper(new FluidStack(fluid, 1000)); // 1000 = 1 bucket in Forge
    }
    
    @Override
    public FluidVariant of(Fluid fluid, CompoundTag nbt) {
        FluidStack fluidStack = new FluidStack(fluid, 1000);
        if (nbt != null) {
            fluidStack.setTag(nbt);
        }
        return new FluidStackWrapper(fluidStack);
    }
    
    public static FluidVariant fromFluidStack(FluidStack fluidStack) {
        return new FluidStackWrapper(fluidStack);
    }
    
    /**
     * Wrapper class that extends our FluidVariant abstract class using Forge's FluidStack.
     */
    public static class FluidStackWrapper extends FluidVariant {
        private final FluidStack fluidStack;
        
        public FluidStackWrapper(FluidStack fluidStack) {
            this.fluidStack = fluidStack;
        }
        
        @Override
        public Fluid getFluid() {
            return fluidStack.getFluid();
        }
        
        @Override
        public CompoundTag getNbt() {
            return fluidStack.getTag();
        }
        
        public FluidStack getFluidStack() {
            return fluidStack;
        }
    }
}

File: forge/src/main/java/dev/kineticcat/hexportation/forge/transfer/FluidVariantImpl.kt
package dev.kineticcat.hexportation.forge.transfer

import dev.kineticcat.hexportation.api.transfer.FluidVariant
import net.minecraft.nbt.CompoundTag
import net.minecraft.world.level.material.Fluid
import net.minecraftforge.fluids.FluidStack

/**
 * Forge implementation using singleton injection pattern.
 */
class FluidVariantImpl : FluidVariant.Implementation() {
    
    companion object {
        @JvmField
        val INSTANCE = FluidVariantImpl()
        
        @JvmStatic
        fun fromFluidStack(fluidStack: FluidStack): FluidVariant = FluidStackWrapper(fluidStack)
    }
    
    override fun of(fluid: Fluid): FluidVariant = 
        FluidStackWrapper(FluidStack(fluid, 1000)) // 1000 = 1 bucket in Forge
    
    override fun of(fluid: Fluid, nbt: CompoundTag?): FluidVariant {
        val fluidStack = FluidStack(fluid, 1000)
        nbt?.let { fluidStack.tag = it }
        return FluidStackWrapper(fluidStack)
    }
    
    /**
     * Wrapper class that extends our FluidVariant abstract class using Forge's FluidStack.
     */
    class FluidStackWrapper(private val fluidStack: FluidStack) : FluidVariant() {
        
        override val fluid: Fluid get() = fluidStack.fluid
        
        override val nbt: CompoundTag? get() = fluidStack.tag
        
        fun getFluidStack() = fluidStack
    }
}